Predicate paths

floriankramer opened the pull request:
Adds support for property paths, but does not support transitive operations (e.g <is-a>*) yet.

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -22,21 +23,116 @@
-QueryPlanner::QueryPlanner(QueryExecutionContext* qec) : _qec(qec) {}
+QueryPlanner::QueryPlanner(QueryExecutionContext* qec)
+    : _qec(qec), _randVarCount(0) {
+  _randomSeed = time(NULL);
```
Do we want to set this to something static to make runs comparable?

floriankramer replied:
Seems like a good idea. I've set the seed to 42.

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -183,70 +264,37 @@ QueryExecutionTree QueryPlanner::createExecutionTree(ParsedQuery& pq) const {
+      for (size_t i = 1; i < lastRow.size(); ++i) {
+        size_t thisCost = lastRow[i].getCostEstimate();
+        if (thisCost < minCost) {
+          minCost = lastRow[i].getCostEstimate();
```
`minCost = thisCost;`

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -672,24 +722,25 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::getOrderByRow(
     // complex than need be.
-    if (t._p == CONTAINS_WORD_PREDICATE || t._p == CONTAINS_ENTITY_PREDICATE) {
-      contextVars.insert(t._s);
+    if (t._p._iri == CONTAINS_WORD_PREDICATE ||
```
Is the above TODO still valid? This doesn't look too complicated

floriankramer replied:
I'm not sure. Do we know who added this TODO?

niklas88 replied:
According to `git blame` it was added by Björn 2 years ago, so I would say that's a no

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -790,245 +840,476 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::seedWithScansAndText(
+}
+
+// _____________________________________________________________________________
+std::string QueryPlanner::generateRandomVarName() {
```
How do we guarantee these don't collide?

niklas88 replied:
Couldn't we just use some (possibly illegal) prefix + a count?

floriankramer replied:
Using a fixed instead of a random prefix seems like a good idea. I'd use a colon though, as it is not allowed as the first character of a variable (only [a-z][A-Z][0-9] '_' and some unicode characters are allowed) and it's a less obscure character.

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -790,245 +840,476 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::seedWithScansAndText(
+  for (ParsedQuery::GraphPattern& child : childPlans) {
+    p._children.insert(p._children.end(), child._children.begin(),
+                       child._children.end());
+    // Ensure the child does not try to delete the pointers
```
Maybe we should really change these child pointers to `std::shared_ptr` but I realize this is a large change, you could add a TODO

floriankramer replied:
Seems like a good idea to me. I've added a TODO for now.

niklas88 commented:
By the way, you can use the following bash script with my ubuntu-clang-format container to format the code on Arch
```
#!/usr/bin/env sh
docker run --rm -v "$(pwd):/input" -u $(id -u):$(id -g) niklas88/ubuntu-clang-format $@
```
With `zsh` I usually run the following command where `ubuntu-clang-format` is the above script in my PATH.
```
ubuntu-clang-format -i (src|test)/**/*.(cpp|h)
```

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -790,245 +840,476 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::seedWithScansAndText(
+      v += 'a' - 26;
+    } else {
+      // Characters A-Z
+      v += 'A';
```
Somehow the names in my test don't look very random e.g. `?impOiAAAAAAAAAAAAAAAAAAAAAAAAAAA2` and `?JHzegEAAAAAAAAAAAAAAAAAAAAAAAAAA1`

floriankramer commented on src/parser/PropertyPathParser.cpp:
```diff
@@ -0,0 +1,279 @@
+  DELIMITER_CHARS['^'] = true;
+
+  // Init the valid characters with all delimiters
+  std::memcpy(VALID_CHARS.data(), DELIMITER_CHARS.data(), 256);
```
I replaced the memcpy with an assignment instead, which should have exactly the same effect.

floriankramer commented on test/SparqlParserTest.cpp:
```diff
@@ -515,6 +519,38 @@ TEST(ParserTest, testParseLiteral) {
   ASSERT_TRUE(caught_exception);
 }
 
+TEST(ParserTest, propertyPaths) {
```
Didn't get to e2e tests yet. I'll add some before this pr is fully finished.

floriankramer commented:
My plan was to implement the transitive operation and add some e2e tests before merging this. Should be done by the end of the week.

niklas88 commented:
The transitive operation doesn't work right in my limited testing (its running on Wikidata Full at the moment). Which I think should have been caught by better E2E Tests, I do realize though that this is somewhat hindered by the limits of the Scientists KB.

The following query gives fewer results than the same but without `/wdt:P279*` which can't be since `*` should also match the empty extension.
```
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
SELECT ?river WHERE
{
  ?river wdt:P31/wdt:P279* wd:Q355304 .
  ?river wdt:P30 wd:Q51 .
}
```
Also the following query adding `OPTIONAL` crashes the instance
```
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
SELECT ?river ?location WHERE
{
  ?river wdt:P31/wdt:P279* wd:Q355304 .
  ?river wdt:P30 wd:Q51 .
  OPTIONAL { ?river wdt:P625 ?location . }
}
```
This can't entirely be a problem with the `*` since the following also crashes
```
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
SELECT ?river ?location
WHERE
{
  ?river wdt:P31/wdt:P279/wdt:P279 wd:Q355304 .
  ?river wdt:P30 wd:Q51 .
  OPTIONAL { ?river wdt:P625 ?location . }
}
```
Interestingly using `wdt:P31|(wdt:P31/wdt:P279)` seems to work and correctly includes the result from both sides of the `|`

floriankramer commented:
I just split the swicth of GraphPatterns to using std::shared_ptrs pretty much everywhere into a separate commit (it was so far part of the single commit in #245) and added it onto this pr. This fixed several memory related issues with the GraphPatternOperations and may fix the two crashes you had.

floriankramer commented:
@niklas88 Currently the Transitive operation with a minimum distance of 0 just adds a connection from everything on the left side of its subquery to itself, which only works in a few highly specific cases. Solving this properly would probably require modification of the property paths before optimization to ensure that all transitive operators with a minimum distance of 0 are replaced by a union over the transitive operator with a minimum distance of 1 and the same property path with the transitive operation removed.

niklas88 commented:
@floriankramer sadly the `shared_ptr<>` change doesn't fix the crashes I reported earlier. Still I think this is a worthwhile change and I'm not surprised that it fixes other memory issues.

floriankramer commented:
@niklas88 With the latest changes the transitive operation with a minimum path length of 0 should work correctly, and I wasn't able to reproduce the crashes. An instance with the latest code is currently running on galera on port 7004.

floriankramer commented on src/engine/QueryPlanner.cpp:
```diff
@@ -1189,50 +1204,89 @@ std::shared_ptr<ParsedQuery::GraphPattern> QueryPlanner::seedFromSequence(
+  // We need a union over every combination of joins that exclude any subset the
+  // child paths which are marked as can_be_null.
+  std::vector<std::shared_ptr<ParsedQuery::GraphPattern>> join_patterns;
+  join_patterns.reserve((1 << num_null_children));
```
I've changed the code slightly to reduce the exponent and I've also limited the number of children to 10 (we can probably safely do more, but in most use cases that have the user input a query more should hopefully not be required). Before resolving empty nodes into a set of unions the sequence path's children are now split into chunks, with each chunk containing at least one non empty node (and thus the chunk itself being non empty).

floriankramer commented on src/engine/TransitivePath.cpp:
```diff
@@ -0,0 +1,234 @@
+  for (size_t i = 0; i < indent; ++i) {
+    os << " ";
+  }
+  os << "TRANSITIVE left " << _left << " right " << _right << " minDist "
```
I've changed it ot `TRANSITIVE_PATH`

niklas88 replied:
I meant they use the same casing as the class name

floriankramer commented on src/engine/IdTable.h:
```diff
@@ -508,7 +508,7 @@ class IdTableStatic : private IdTableImpl<COLS> {
   using ConstRow = typename IdTableImpl<COLS>::const_row_type;
   using iterator = typename IdTableImpl<COLS>::iterator;
-  using const_iterator = const typename IdTableImpl<COLS>::iterator;
+  using const_iterator = typename IdTableImpl<COLS>::iterator;
```
In theory we should have a const_iterator that does not allow changed values, but when I began implementing that I didn't find a nice way to write such an iterator that does not lead to insane amounts of code duplication (as the only difference is removing some const versions of methods) and the templating with base classes gets annoying quickly, as the iterator iterator class is inside of another class.

niklas88 replied:
Yeah I think it's ok to have it not actively enforce the constness

floriankramer commented on src/engine/QueryPlanner.cpp:
```diff
@@ -107,26 +184,73 @@ QueryExecutionTree QueryPlanner::createExecutionTree(ParsedQuery& pq) const {
-              new Union(_qec, left->_qet, right->_qet));
-          QueryExecutionTree& tree = *unionPlans.back()._qet.get();
+              std::make_shared<Union>(_qec, left->_qet, right->_qet));
+          QueryExecutionTree& tree = *childPlanStorage.back()._qet.get();
```
childPlanStorage stores `SubtreePlans`. The reference instead of a shared_ptr is a microoptimization, I guess, as we know that the shared_ptr will not be deleted and the storage is not freed.  `unionPlans` was renamed to reflect that it also stores `SubtreePlans` for `TransitivePath` operations (instead of just storing them for `Union` operations).

floriankramer commented on src/engine/QueryPlanner.cpp:
```diff
@@ -737,11 +840,11 @@ QueryPlanner::TripleGraph QueryPlanner::createTripleGraph(
 vector<QueryPlanner::SubtreePlan> QueryPlanner::seedWithScansAndText(
     const QueryPlanner::TripleGraph& tg,
-    const vector<QueryPlanner::SubtreePlan*>& children) const {
+    const vector<const QueryPlanner::SubtreePlan*>& children) {
```
Currently we don't use `shared_ptr` as the data is stored in the `childPlanStorage` and `subqueryPlans` vectors which are guaranteed to have the required lifetime to ensure that the data is never used after deletion. Unless we want to use these `SubtreePlans` somewhere else `shared_ptr` seems like an unecessary overhead to me (although it might not be noticeable in the end, given the small size of the `children` vector).

niklas88 commented on src/engine/QueryPlanner.cpp:
```diff
@@ -790,245 +888,528 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::seedWithScansAndText(
+  op->_pathData._left = left;
+  op->_pathData._right = right;
+  op->_pathData._min = 1;
+  op->_pathData._max = std::numeric_limits<size_t>::max();
```
Ok yes that makes sense.

niklas88 commented:
@floriankramer are you getting mails for failing CI or is that still not working?

floriankramer commented:
@niklas88 I didn't see a mail for the travis failure. Do I have to subscribe to those somewhere?

niklas88 commented:
@floriankramer I would guess that this is controlled by the GitHub notification settings in your profile. I'm not sure however since Checks aren't explicitly mentioned there.

niklas88 commented:
@floriankramer I'm currently testing this on Wikidata Full (together with #250).

One thing I've found that was broken before but might be a good fit for fixing in this PR is that a query with the same variable on both sides of a predicate crashes QLever. Here is a Scientists example query
```
SELECT ?x WHERE {
   ?x <is-a> ?x
}
```
In particular this also happens for the following useful [Wikidata Full example](https://www.wikidata.org/wiki/Wikidata:SPARQL_query_service/queries/examples#Administrative_entities_located_within_themselves)
```
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
SELECT ?x WHERE {
   ?x wdt:P131 ?x .
}
```
I'll investigate but if you already know what the problem is all the better.

niklas88 commented on src/parser/ParsedQuery.cpp:
```diff
@@ -394,10 +521,43 @@ void ParsedQuery::GraphPattern::toString(std::ostringstream& os,
+void ParsedQuery::GraphPattern::recomputeIds(size_t* id_count) {
+  bool allocatedIdCounter = false;
+  if (id_count == nullptr) {
+    id_count = new size_t(0);
```
I missed this during the review but we could just use a stack var here and pass a pointer to that to the recursive call only if the parameter was null. This leaves the same variable in the recursive calls unused but that should still be faster than an allocation.
